# Data Quality food_access_research_atlas

In [1]:
import pandas as pd

In [2]:
#reading in food access data domain 
food_access_2015 = pd.read_csv("food_access_research_atlas_2015.csv")
food_access_2015.head()

CensusTract    State   County  Urban  POP2010  OHU2010  GroupQuartersFlag  \
0   1001020100  Alabama  Autauga      1     1912      693                  0   
1   1001020200  Alabama  Autauga      1     2170      743                  0   
2   1001020300  Alabama  Autauga      1     3373     1256                  0   
3   1001020400  Alabama  Autauga      1     4386     1722                  0   
4   1001020500  Alabama  Autauga      1    10766     4082                  0   

   NUMGQTRS  PCTGQTRS  LILATracts_1And10  ...  TractSeniors  TractWhite  \
0         0  0.000000                  0  ...           221        1622   
1       181  0.083410                  0  ...           214         888   
2         0  0.000000                  0  ...           439        2576   
3         0  0.000000                  0  ...           904        4086   
4       181  0.016812                  0  ...          1126        8666   

   TractBlack  TractAsian  TractNHOPI  TractAIAN  TractOMultir  TractHispanic  \
0         217          14           0         14            45             44   
1        1217           5           0          5            55             75   
2         647          17           5         11           117             87   
3         193          18           4         11            74             85   
4        1437         296           9         48           310            355   

   TractHUNV  TractSNAP  
0         26        112  
1         87        202  
2        108        120  
3         19         82  
4        198        488  

[5 rows x 147 columns]

In [3]:
#Keep just the state of utah data
utah_mask = food_access_2015["State"]  == 'Utah'
food_access_2015 = food_access_2015.loc[utah_mask]
food_access_2015

CensusTract State     County  Urban  POP2010  OHU2010  \
66726  49001100100  Utah     Beaver      0     3818     1338   
66727  49001100200  Utah     Beaver      0     2811      927   
66728  49003960100  Utah  Box Elder      0     3152      978   
66729  49003960200  Utah  Box Elder      0     6901     2186   
66730  49003960300  Utah  Box Elder      1     8186     2558   
...            ...   ...        ...    ...      ...      ...   
67309  49057210900  Utah      Weber      1     6316     2268   
67310  49057211000  Utah      Weber      1     3073     1089   
67311  49057211100  Utah      Weber      1     5785     2093   
67312  49057211201  Utah      Weber      1     4452     1604   
67313  49057211202  Utah      Weber      1     6531     2560   

       GroupQuartersFlag  NUMGQTRS  PCTGQTRS  LILATracts_1And10  ...  \
66726                  0        14  0.003667                  0  ...   
66727                  0        10  0.003557                  0  ...   
66728                  0         9  0.002855                  0  ...   
66729                  0         0  0.000000                  0  ...   
66730                  0        32  0.003909                  1  ...   
...                  ...       ...       ...                ...  ...   
67309                  0         0  0.000000                  0  ...   
67310                  0         0  0.000000                  1  ...   
67311                  0       136  0.023509                  1  ...   
67312                  0        50  0.011231                  0  ...   
67313                  0       108  0.016536                  0  ...   

       TractSeniors  TractWhite  TractBlack  TractAsian  TractNHOPI  \
66726           539        3409           5          54          11   
66727           293        2491          11          16           8   
66728           279        2981           6           8           4   
66729           778        6475          16          75          18   
66730           717        7407          29         120          15   
...             ...         ...         ...         ...         ...   
67309          1033        5691          87          79          24   
67310           429        2756          67          16           4   
67311           784        4918         120         100          27   
67312           609        4114          38          68           9   
67313          1148        6001          71         116          26   

       TractAIAN  TractOMultir  TractHispanic  TractHUNV  TractSNAP  
66726         43           296            414         26        105  
66727         33           252            302         19         14  
66728          8           145            183         18         51  
66729         22           295            406         37        134  
66730         41           574            927        233        431  
...          ...           ...            ...        ...        ...  
67309         31           404            654         73        148  
67310         24           206            321         33         95  
67311         42           578            840        279        296  
67312         28           195            294        127         76  
67313         21           296            394         63        120  

[588 rows x 147 columns]

In [5]:
food_access_2015.to_csv('food_access_2015.csv', index=False)

In [22]:
#keeping the columns I want
#Field	LongName
#CensusTract	Census tract
#State	State
#County	County
#Urban	Urban tract
#POP2010	Population, tract total
#PovertyRate	Tract poverty rate
#MedianFamilyIncome	Tract median family income
#LA1and10	Low access tract at 1 mile for urban areas and 10 miles for rural areas
#LAhalfand10	Low access tract at 1/2 mile for urban areas and 10 miles for rural areas
#LA1and20	Low access tract at 1 mile for urban areas and 20 miles for rural areas
#LATracts_half	Low access tract at 1/2 mile
#LATracts1	Low access tract at 1 mile
#LATracts10	Low access tract at 10 miles
#LATracts20	Low access tract at 20 miles
#TractWhite	Tract White population, number
#TractBlack	Tract Black or African American population, number
#TractAsian	Tract Asian population, number
#TractNHOPI	Tract Native Hawaiian and Other Pacific Islander population, number
#TractAIAN	Tract American Indian and Alaska Native population, number
#TractOMultir	Tract Other/Multiple race population, number
#TractHispanic	Tract Hispanic or Latino population, number

food_access_2015 = food_access_2015[['CensusTract', 'State', 'County', 'Urban', 'POP2010', 'PovertyRate',
                                'MedianFamilyIncome', 'LA1and10', 'LAhalfand10', 'LA1and20', 
                                'LATracts_half', 'LATracts1', 'LATracts10', 'LATracts20',
                                'TractWhite', 'TractBlack', 'TractAsian', 
                                'TractNHOPI', 'TractAIAN', 'TractOMultir', 'TractHispanic']]
food_access_2015.head()

CensusTract State     County  Urban  POP2010  PovertyRate  \
66726  49001100100  Utah     Beaver      0     3818         15.1   
66727  49001100200  Utah     Beaver      0     2811         10.5   
66728  49003960100  Utah  Box Elder      0     3152         13.7   
66729  49003960200  Utah  Box Elder      0     6901          9.2   
66730  49003960300  Utah  Box Elder      1     8186         14.7   

       MedianFamilyIncome  LA1and10  LAhalfand10  LA1and20  ...  LATracts1  \
66726               56250         0            0         0  ...          0   
66727               62972         1            1         0  ...          0   
66728               62798         1            1         1  ...          0   
66729               63638         0            0         0  ...          0   
66730               56216         1            1         1  ...          1   

       LATracts10  LATracts20  TractWhite  TractBlack  TractAsian  TractNHOPI  \
66726           0           0        3409           5          54          11   
66727           1           0        2491          11          16           8   
66728           1           1        2981           6           8           4   
66729           0           0        6475          16          75          18   
66730           0           0        7407          29         120          15   

       TractAIAN  TractOMultir  TractHispanic  
66726         43           296            414  
66727         33           252            302  
66728          8           145            183  
66729         22           295            406  
66730         41           574            927  

[5 rows x 21 columns]

# Attribute domain constraints

Basic aggregate statistics for the food_access_2015 data domain 

The descriptive statistics seem reasonable. The Low access columns and urban all have a min of 0 and Max of 1 as expected since they are binary results. All other columns seem reasonable. 

In [24]:
food_access_2015.describe()

CensusTract       Urban       POP2010  PovertyRate  \
count  5.880000e+02  588.000000    588.000000   588.000000   
mean   4.903629e+10    0.846939   4700.484694    13.472619   
std    1.535444e+07    0.360353   2073.815323    11.281999   
min    4.900110e+10    0.000000      0.000000     0.000000   
25%    4.903510e+10    1.000000   3438.500000     5.900000   
50%    4.903511e+10    1.000000   4442.000000    10.150000   
75%    4.904900e+10    1.000000   5785.750000    17.700000   
max    4.905721e+10    1.000000  21591.000000    85.500000   

       MedianFamilyIncome    LA1and10  LAhalfand10    LA1and20  LATracts_half  \
count          588.000000  588.000000   588.000000  588.000000     588.000000   
mean         69030.653061    0.447279     0.828231    0.423469       0.778912   
std          24620.559447    0.497636     0.377500    0.494529       0.415333   
min              0.000000    0.000000     0.000000    0.000000       0.000000   
25%          53303.500000    0.000000     1.000000    0.000000       1.000000   
50%          66166.500000    0.000000     1.000000    0.000000       1.000000   
75%          83373.750000    1.000000     1.000000    1.000000       1.000000   
max         196750.000000    1.000000     1.000000    1.000000       1.000000   

        LATracts1  LATracts10  LATracts20    TractWhite  TractBlack  \
count  588.000000  588.000000  588.000000    588.000000  588.000000   
mean     0.397959    0.049320    0.025510   4046.870748   49.807823   
std      0.489894    0.216719    0.157803   1900.595299   56.328576   
min      0.000000    0.000000    0.000000      0.000000    0.000000   
25%      0.000000    0.000000    0.000000   2866.750000   16.000000   
50%      0.000000    0.000000    0.000000   3777.000000   32.000000   
75%      1.000000    0.000000    0.000000   4933.000000   64.000000   
max      1.000000    1.000000    1.000000  20051.000000  497.000000   

        TractAsian  TractNHOPI    TractAIAN  TractOMultir  TractHispanic  
count   588.000000  588.000000   588.000000    588.000000     588.000000  
mean     94.022109   41.758503    55.998299    412.027211     609.421769  
std     100.750782   61.646292   207.615853    397.855190     606.105033  
min       0.000000    0.000000     0.000000      0.000000       0.000000  
25%      30.000000    9.000000    15.000000    167.750000     228.750000  
50%      67.000000   21.000000    28.500000    271.000000     384.000000  
75%     123.000000   46.250000    52.000000    511.000000     785.000000  
max    1064.000000  563.000000  3771.000000   2467.000000    3496.000000

In [25]:
import matplotlib
import matplotlib.pyplot as plt
plt.style.use('ggplot')
# This next line tells jupyter to render the images inline
%matplotlib inline

The value counts for urban show that there are more urban census tracts than rural in the state of utah. LAhalfand10 is the varaible for low access tract at 1/2 mile for urban areas and 10 miles for rural areas. There are more 1's I assume mean yes than 0's which I assume mean no. There are 585 census tracts. Each row in the dataset represents a unique census tract. Salt Lake city county has the most census tracts with 211. 

In [26]:
food_access_2015["Urban"].value_counts()

Urban
1    498
0     90
Name: count, dtype: int64

In [27]:
food_access_2015["LAhalfand10"].value_counts()

LAhalfand10
1    487
0    101
Name: count, dtype: int64

In [28]:
food_access_2015.groupby(['County']).describe()
#2015 data has 3 additional census tracts one additional in salt lake county, utah county, and tooele county 

CensusTract                                             \
                 count          mean            std           min   
County                                                              
Beaver             2.0  4.900110e+10      70.710678  4.900110e+10   
Box Elder         11.0  4.900396e+10     240.757970  4.900396e+10   
Cache             26.0  4.900504e+10  192064.627410  4.900500e+10   
Carbon             5.0  4.900700e+10     207.364414  4.900700e+10   
Daggett            1.0  4.900996e+10            NaN  4.900996e+10   
Davis             54.0  4.901113e+10     568.639313  4.901113e+10   
Duchesne           3.0  4.901394e+10     152.752523  4.901394e+10   
Emery              3.0  4.901598e+10     152.752523  4.901598e+10   
Garfield           2.0  4.901700e+10      70.710678  4.901700e+10   
Grand              2.0  4.901900e+10      70.710678  4.901900e+10   
Iron               8.0  4.902111e+10     226.882246  4.902111e+10   
Juab               2.0  4.902301e+10      70.710678  4.902301e+10   
Kane               2.0  4.902513e+10      70.710678  4.902513e+10   
Millard            3.0  4.902797e+10     100.000000  4.902797e+10   
Morgan             2.0  4.902997e+10      70.710678  4.902997e+10   
Piute              1.0  4.903196e+10            NaN  4.903196e+10   
Rich               1.0  4.903395e+10            NaN  4.903395e+10   
Salt Lake        212.0  4.903511e+10   59886.844401  4.903510e+10   
San Juan           4.0  4.903796e+10   20842.424683  4.903794e+10   
Sanpete            5.0  4.903997e+10     158.113883  4.903997e+10   
Sevier             5.0  4.904198e+10     158.113883  4.904198e+10   
Summit            13.0  4.904396e+10      96.823723  4.904396e+10   
Tooele            11.0  4.904521e+10  256022.126919  4.904513e+10   
Uintah             6.0  4.904796e+10   11480.270637  4.904794e+10   
Utah             128.0  4.904901e+10   86335.116641  4.904900e+10   
Wasatch            4.0  4.905196e+10    9867.454923  4.905194e+10   
Washington        21.0  4.905327e+10     520.621501  4.905327e+10   
Wayne              1.0  4.905598e+10            NaN  4.905598e+10   
Weber             50.0  4.905721e+10    4824.614888  4.905720e+10   

                                                                    Urban  \
                     25%           50%           75%           max  count   
County                                                                      
Beaver      4.900110e+10  4.900110e+10  4.900110e+10  4.900110e+10    2.0   
Box Elder   4.900396e+10  4.900396e+10  4.900396e+10  4.900396e+10   11.0   
Cache       4.900500e+10  4.900500e+10  4.900500e+10  4.900598e+10   26.0   
Carbon      4.900700e+10  4.900700e+10  4.900700e+10  4.900700e+10    5.0   
Daggett     4.900996e+10  4.900996e+10  4.900996e+10  4.900996e+10    1.0   
Davis       4.901113e+10  4.901113e+10  4.901113e+10  4.901113e+10   54.0   
Duchesne    4.901394e+10  4.901394e+10  4.901394e+10  4.901394e+10    3.0   
Emery       4.901598e+10  4.901598e+10  4.901598e+10  4.901598e+10    3.0   
Garfield    4.901700e+10  4.901700e+10  4.901700e+10  4.901700e+10    2.0   
Grand       4.901900e+10  4.901900e+10  4.901900e+10  4.901900e+10    2.0   
Iron        4.902111e+10  4.902111e+10  4.902111e+10  4.902111e+10    8.0   
Juab        4.902301e+10  4.902301e+10  4.902301e+10  4.902301e+10    2.0   
Kane        4.902513e+10  4.902513e+10  4.902513e+10  4.902513e+10    2.0   
Millard     4.902797e+10  4.902797e+10  4.902797e+10  4.902797e+10    3.0   
Morgan      4.902997e+10  4.902997e+10  4.902997e+10  4.902997e+10    2.0   
Piute       4.903196e+10  4.903196e+10  4.903196e+10  4.903196e+10    1.0   
Rich        4.903395e+10  4.903395e+10  4.903395e+10  4.903395e+10    1.0   
Salt Lake   4.903511e+10  4.903511e+10  4.903511e+10  4.903598e+10  212.0   
San Juan    4.903794e+10  4.903796e+10  4.903798e+10  4.903798e+10    4.0   
Sanpete     4.903997e+10  4.903997e+10  4.903997e+10  4.903997e+10    5.0   
Sevier      4.904198e+10  4.904

In [29]:
food_access_2015['CensusTract'].value_counts()

CensusTract
49001100100    1
49049000707    1
49049000203    1
49049000204    1
49049000205    1
              ..
49035111304    1
49035111305    1
49035111306    1
49035111400    1
49057211202    1
Name: count, Length: 588, dtype: int64

In [30]:
food_access_2015["County"].value_counts()

County
Salt Lake     212
Utah          128
Davis          54
Weber          50
Cache          26
Washington     21
Summit         13
Tooele         11
Box Elder      11
Iron            8
Uintah          6
Carbon          5
Sanpete         5
Sevier          5
Wasatch         4
San Juan        4
Duchesne        3
Emery           3
Millard         3
Beaver          2
Kane            2
Juab            2
Grand           2
Garfield        2
Morgan          2
Rich            1
Piute           1
Daggett         1
Wayne           1
Name: count, dtype: int64

Optionality was described as number of missing values in the paper. There are no missing values. 


In [32]:
#checking for null values
food_access_2015.isnull().sum()

CensusTract           0
State                 0
County                0
Urban                 0
POP2010               0
PovertyRate           0
MedianFamilyIncome    0
LA1and10              0
LAhalfand10           0
LA1and20              0
LATracts_half         0
LATracts1             0
LATracts10            0
LATracts20            0
TractWhite            0
TractBlack            0
TractAsian            0
TractNHOPI            0
TractAIAN             0
TractOMultir          0
TractHispanic         0
dtype: int64

# Relational integrity rules

The paper refers to checking for identity. In this data set that means making sure the CensusTract are unique. Using food_access_2015["CensusTract"].describe() shows that there are 588 rows in the table and 588 unique census tracts. 

In [35]:
food_access_2015["CensusTract"].describe()

count    5.880000e+02
mean     4.903629e+10
std      1.535444e+07
min      4.900110e+10
25%      4.903510e+10
50%      4.903511e+10
75%      4.904900e+10
max      4.905721e+10
Name: CensusTract, dtype: float64

The paper refers to checking for reference. Will need to check that the census tract can be used to join the individual datasets together. 

# Historical data rules 

Historical data rules according to the paper are rules that look at "time-varying data". Data sources: In the 2017 report, a directory of supermarkets, supercenters, and large grocery
stores within the United States, including Alaska and Hawaii, was derived from merging the
2015 STARS directory of stores authorized to accept Supplemental Nutrition Assistance
Program (SNAP) benefits and the 2015 Trade Dimensions TDLinx directory of stores. According to the documentation included with the data download. 

# State dependent objects rules

The values for urban and for the low access columns can only be 0 or 1 not both or a value in between. Looking at the values counts in the attribute doman constraint section shows that this is the the case. The value counts show that 0 and 1 are the only values in the column and the totals add up to 588 as expected. 

# Attribute dependency rules

To assess attribute dependency rules we should test that redundant attributes in different data sources have identical values. 

The sum of the races equals the the total population for the census tract. 

In [37]:
#Derived attributes: aggregated valu must equal the total of the atmoic level values.
#Checking if the sum of the race categories equals the total population
food_access_2015['total_race_count'] = food_access_2015['TractWhite'] + food_access_2015['TractBlack'] + food_access_2015['TractAsian'] +  food_access_2015['TractNHOPI'] + food_access_2015['TractAIAN'] + food_access_2015['TractOMultir']

food_access_2015[['POP2010','total_race_count']]

POP2010  total_race_count
66726     3818              3818
66727     2811              2811
66728     3152              3152
66729     6901              6901
66730     8186              8186
...        ...               ...
67309     6316              6316
67310     3073              3073
67311     5785              5785
67312     4452              4452
67313     6531              6531

[588 rows x 2 columns]

# Conclusion

Overall, the quality of data is good in the the food access 2015 data set. There are no  missing values. The outputs in the columns are as expected for example they are a 0 or a 1 when they should be and the sum of the aggregrate race data matches the total population data. More assessment should be done when we combine the datasets to make sure the values are the same across sources for redundant columns. 

The facets dataset used the Low access to healthy food (1/2 mile). Since Utah has both rural and urban census tracts I think we should use the Low access tract at 1/2 mile for urban areas and 10 miles for rural areas column. That is the column I focused on in this assessment. There are other options to explore. 



# Resources

Description of the columns: https://www.ers.usda.gov/data-products/food-access-research-atlas/documentation/#distance

Download the data: https://www.ers.usda.gov/data-products/food-access-research-atlas/download-the-data/ in the zip files there is one spreadsheet that gives a brief explanation of the columns and another excel file with the actual data

See the right hand side of both the links above for additional information about this data source. 
